### PR TITLE
pkg: archive: don't fail Untar if xattrs are not supported

### PR DESCRIPTION
**Pull request by cyphar**
Since certain filesystems don't support extended attributes, ignore
errors produced (emitting a warning) when attempting to apply extended
attributes to file.

This is a **preliminary** fix for #21216. However, a full fix requires
changing `ADD` and `COPY` to use `ApplyDiff`in the layer store. If this
isn't a "good enough" fix, then I'll get to work on the `ApplyDiff` fix.

excited kitteh

Signed-off-by: Aleksa Sarai asarai@suse.de

/cc @tianon @thaJeztah

### CODE REVIEW

**Comment by cyphar** [comment]
@tianon I should probably make it clear that the Docker Hub official images builder should switch to a more modern graphdriver (`btrfs` and `devicemapper` are good choices IMO). This PR causes AUFS to strip `xattrs` (which is what happens with `docker import` at the moment as well).

**Comment by tianon** [comment]
Yeah, I was afraid that might be the case.  Currently, AUFS is the most well-supported backend, and the only real issue we've run into (in the context of building the official images) with it is extended attributes, which is more than I can say for either BTRFS _or_ overlayfs (and devicemapper is such a PITA to configure appropriately that I've honestly never seen a well-tuned installation of it in the real world).

**Comment by cyphar** [comment]
@tianon My understanding was that Debian and Ubuntu were planning to drop support for AUFS (since it's unlikely it'll be merged into the mainline kernel). My perspective has been that almost all of the weird builder and graphdriver bugs (`RUN chmod`, layer limitations, etc) have all been caused by AUFS. Maybe I'm not seeing the full story, but I've seen far fewer issues that have been a problem with the *kernel* when using `btrfs` and `devicemapper` (but I agree `devicemapper` is probably the hardest thing to set up properly).

I do agree that overlayfs has quite a few issues that need to be fixed on the kernel side (and there seems to be a lack of responsiveness from upstream too, as well as blatant POSIX non-compliance issues). But at SUSE we've been shipping `btrfs` (as the default filesystem) with SUSE-based distros for quite a while, and it seems to be working well for us.

And lastly, I personally feel quite iffy about the fact that the "most well-supported backend" is using an out-of-tree patchset for the Linux kernel. Maybe we could provide more love to the other backends so we can ween ourselves off the first backend we ever supported? We did it to LXC, we can do it to AUFS too. :P

File: pkg/archive/archive.go [hunk 1]
```diff
@@ -426,8 +426,14 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 	}
 
 	for key, value := range hdr.Xattrs {
+		hadError := false
 		if err := system.Lsetxattr(path, key, []byte(value), 0); err != nil {
-			return err
+			// We ignore errors here because not all graphdrivers support xattrs.
```
Comment (calavera) [inline pkg/archive/archive.go]: why don't you write the warning here directly?

**Comment by calavera** [comment]
> Currently, AUFS is the most well-supported backend

I doubt it's the most used anymore, the majority of users fall back to devicemapper.

Anyways, that's orthogonal to this patch. I agree that we should not fail when xattrs are not supported.

LGTM

File: pkg/archive/archive.go (hunk 1 above)
Comment (cyphar) [inline pkg/archive/archive.go]: It would spam the log. I can add it back if you like (or I can aggregate the key-value mappings into one error).

Comment (calavera) [inline pkg/archive/archive.go]: I think aggregating into one error is more consistent with other places where we do that.

**Comment by cyphar** [comment]
There, I've fixed up how it outputs the errors.

/ping @calavera @tianon

**Comment by tonistiigi** [comment]
LGTM

**Comment by tianon** [comment]
LGTM

**Comment by cyphar** [comment]
NOTE: I still think in order to fix this properly, we should be using `ApplyDiff` in the layer store when doing `ADD` and `COPY`.

**Comment by rhatdan** [comment]
Any reason this patch is not checking ENOTSUPP?  I think it should only ignore that. Other potential errors like EPERM or ENOSPC should probably not be ignored.

**Comment by runcom** [comment]
@cyphar ^^

**Comment by cyphar** [comment]
I want to figure out why `docker import` works while `ADD` doesn't -- my original "full take" PR was to merge the two together because we currently have two pieces of code that mess with the layer store. But yes, we should probably just be checking `ENOTSUPP`. I can write a PR for this if you like.

**Comment by rhatdan** [comment]
I am rather busy right now, so if you could throw one together, that would be great.

**Comment by cyphar** [comment]
@rhatdan @runcom I've opened #23410 that should fix this. PTAL.

**Pull request merged**